mempool/clist_mempool.go

```diff
@@ -648,41 +712,49 @@ func (mem *CListMempool) recheckTxs() {
 		return
 	}
 
-	mem.recheck.init(mem.txs.Front(), mem.txs.Back())
+	// Recheck all transactions in each lane, sequentially.
+	// TODO: parallelize rechecking on lanes?
+	for _, lane := range mem.sortedLanes {
+		mem.logger.Debug("recheck lane", "height", mem.height.Load(), "lane", lane, "num-txs", mem.lanes[lane].Len())
+		mem.recheck.init(mem.lanes[lane].Front(), mem.lanes[lane].Back())
+
+		// NOTE: handleCheckTxResponse may be called concurrently, but CheckTx cannot be executed concurrently
+		// because this function has the lock (via Update and Lock).
+		for e := mem.lanes[lane].Front(); e != nil; e = e.Next() {
+			tx := e.Value.(*mempoolTx).tx
+			mem.recheck.numPendingTxs.Add(1)
+
+			// Send CheckTx request to the app to re-validate transaction.
+			resReq, err := mem.proxyAppConn.CheckTxAsync(context.TODO(), &abci.CheckTxRequest{
+				Tx:   tx,
+				Type: abci.CHECK_TX_TYPE_RECHECK,
+			})
+			if err != nil {
+				panic(fmt.Errorf("(re-)CheckTx request for tx %s failed: %w", log.NewLazySprintf("%v", tx.Hash()), err))
+			}
+			resReq.SetCallback(mem.handleRecheckTxResponse(tx))
+		}
 
-	// NOTE: CheckTx for new transactions cannot be executed concurrently
-	// because this function has the lock (via Update and Lock).
-	for e := mem.txs.Front(); e != nil; e = e.Next() {
-		tx := e.Value.(*mempoolTx).tx
-		mem.recheck.numPendingTxs.Add(1)
+		// Flush any pending asynchronous recheck requests to process.
+		mem.proxyAppConn.Flush(context.TODO())
 
-		// Send CheckTx request to the app to re-validate transaction.
-		resReq, err := mem.proxyAppConn.CheckTxAsync(context.TODO(), &abci.CheckTxRequest{
-			Tx:   tx,
-			Type: abci.CHECK_TX_TYPE_RECHECK,
-		})
-		if err != nil {
-			panic(fmt.Errorf("(re-)CheckTx request for tx %s failed: %w", log.NewLazySprintf("%v", tx.Hash()), err))
+		// Give some time to finish processing the responses; then finish the rechecking process, even
+		// if not all txs were rechecked.
+		select {
+		case <-time.After(mem.config.RecheckTimeout):
+			mem.recheck.setDone()
```

Comment:
This resets `isRechecked` and sets it to false. It means that for a brief moment, until the next loop iteration begins, the mempool thinks rechecking is done.

Reply:
Fixed with #3862.